File: tutorials/mpi/examples/mpi_scatter.f
C ****************************************************************************
C FILE: mpi_scatter.f
C DESCRIPTION:
C   MPI tutorial example code: Collective Communications
C AUTHOR: Blaise Barney
C LAST REVISED: 04/02/05
C ****************************************************************************

      program scatter
      include 'mpif.h'

      integer SIZE
      parameter(SIZE=4)
      integer numtasks, rank, sendcount, recvcount, source, ierr
      real*4 sendbuf(SIZE,SIZE), recvbuf(SIZE)

C     Fortran stores this array in column major order, so the 
C     scatter will actually scatter columns, not rows.
      data sendbuf /1.0, 2.0, 3.0, 4.0, 
     &              5.0, 6.0, 7.0, 8.0,
     &              9.0, 10.0, 11.0, 12.0, 
     &              13.0, 14.0, 15.0, 16.0 /

      call MPI_INIT(ierr)
      call MPI_COMM_RANK(MPI_COMM_WORLD, rank, ierr)
      call MPI_COMM_SIZE(MPI_COMM_WORLD, numtasks, ierr)

      if (numtasks .eq. SIZE) then
         source = 1
         sendcount = SIZE
         recvcount = SIZE
         call MPI_SCATTER(sendbuf, sendcount, MPI_REAL, recvbuf, 
     &        recvcount, MPI_REAL, source, MPI_COMM_WORLD, ierr)
         print *, 'rank= ',rank,' Results: ',recvbuf 
      else
         print *, 'Must specify',SIZE,' processors.  Terminating.' 
      endif

      call MPI_FINALIZE(ierr)

      end

File: mpi/examples/mpi_scatter.c
/******************************************************************************
* FILE: mpi_scatter.c
* DESCRIPTION:
*   MPI tutorial example code: Collective Communications
* AUTHOR: Blaise Barney
* LAST REVISED: 04/13/05
******************************************************************************/
#include "mpi.h"
#include <stdio.h>
#include <stdlib.h>
#define SIZE 4

int main (int argc, char *argv[])
{
int numtasks, rank, sendcount, recvcount, source;
float sendbuf[SIZE][SIZE] = {
  {1.0, 2.0, 3.0, 4.0},
  {5.0, 6.0, 7.0, 8.0},
  {9.0, 10.0, 11.0, 12.0},
  {13.0, 14.0, 15.0, 16.0}  };
float recvbuf[SIZE];

MPI_Init(&argc,&argv);
MPI_Comm_rank(MPI_COMM_WORLD, &rank);
MPI_Comm_size(MPI_COMM_WORLD, &numtasks);

if (numtasks == SIZE) {
  source = 1;
  sendcount = SIZE;
  recvcount = SIZE;
  MPI_Scatter(sendbuf,sendcount,MPI_FLOAT,recvbuf,recvcount,
             MPI_FLOAT,source,MPI_COMM_WORLD);

  printf("rank= %d  Results: %f %f %f %f\n",rank,recvbuf[0],
         recvbuf[1],recvbuf[2],recvbuf[3]);
  }
else
  printf("Must specify %d processors. Terminating.\n",SIZE);

MPI_Finalize();
}
